src/components/popup/container.tsx

```diff
@@ -1,23 +1,43 @@
-import React, { memo } from 'react';
+import React, { memo, useCallback, useEffect } from 'react';
 import { Flex, color } from '@stacks/ui';
 import { SettingsPopover } from './settings-popover';
 import { useRecoilValue } from 'recoil';
 import { hasRehydratedVaultStore } from '@store/wallet';
 import { Header } from '@components/header';
+import { useTxState } from '@common/hooks/use-tx-state';
+import { useWallet } from '@common/hooks/use-wallet';
 
 interface PopupHomeProps {
   title?: string;
   onClose?: () => void;
   hideActions?: boolean;
+  requestType?: string;
 }
 
 export const PopupContainer: React.FC<PopupHomeProps> = memo(
-  ({ children, title, onClose, hideActions }) => {
+  ({ children, title, onClose, hideActions, requestType }) => {
     const hasRehydratedVault = useRecoilValue(hasRehydratedVaultStore);
+    const { handleCancelTransaction } = useTxState();
+    const { handleCancelAuthentication } = useWallet();
 
     if (!hasRehydratedVault) {
       console.error('No hasRehydratedVault, rendered null');
     }
+
+    const handleUnmount = useCallback(async () => {
+      if (requestType === 'auth') {
+        handleCancelAuthentication();
+      }
+      if (requestType === 'transaction') {
+        await handleCancelTransaction();
+      }
+    }, [requestType, handleCancelAuthentication, handleCancelTransaction]);
+
+    useEffect(() => {
+      window.addEventListener('beforeunload', handleUnmount);
+      return () => window.removeEventListener('beforeunload', handleUnmount);
+    }, [handleUnmount]);
+
```

Comment:
Moved this all into the `PopupContainer` and added a `requestType` prop to handle both types of requests. Not great, but the best I could see how to do without refactoring the messaging.

Reply:
Makes sense. Perhaps a comment would be worthy here.